operators/src/pro/ml/xgboost.rs

```diff
@@ -0,0 +1,694 @@
+            .iter()
+            .map(|init_raster| {
+                let typed_raster_qp = match init_raster.query_processor() {
+                    Ok(r) => r,
+                    Err(e) => {
+                        return Err(Error::XGBoost {
+                            source: XGBoostModuleError::OperatorsError {
+                                source: Box::new(e),
+                            },
+                        })
+                    }
+                };
+                let converted = typed_raster_qp.into_f32();
+                Ok(converted)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(QueryProcessorOut(Box::new(XgboostProcessor::new(
+            vec_of_rqps,
+            self.model_sub_path.clone(),
+            self.no_data_value,
+        ))))
+    }
+}
+
+struct XgboostProcessor<Q, P>
+where
+    Q: RasterQueryProcessor<RasterType = P>,
+{
+    sources: Vec<Q>,
+    model_file: PathBuf,
+    no_data_value: f32,
+}
+
+impl<Q, P> XgboostProcessor<Q, P>
+where
+    Q: RasterQueryProcessor<RasterType = P>,
+    P: Pixel,
+{
+    pub fn new(sources: Vec<Q>, model_file_path: PathBuf, no_data_value: f32) -> Self {
+        Self {
+            sources,
+            model_file: model_file_path,
+            no_data_value,
+        }
+    }
+
+    async fn process_tile_async(
+        &self,
+        bands_of_tile: Result<Vec<RasterTile2D<P>>>,
+        pool: Arc<ThreadPool>,
+    ) -> Result<BaseTile<GridOrEmpty<GridShape<[usize; 2]>, f32>>> {
+        let bot = match bands_of_tile {
+            Ok(b) => b,
+            Err(_) => return Err(XGBoostModuleError::NoInputData.into()),
+        };
+
+        ensure!(!bot.is_empty(), self::error::NoInputData);
+
+        let tile = bot.get(0).context(self::error::BaseTile)?;
+
+        // gather the data
+        let grid_shape = tile.grid_shape();
+        let n_bands = bot.len() as i32;
+        let props = tile.properties.clone(); // = &tile.properties;
+        let time = tile.time;
+        let tile_position = tile.tile_position;
+        let global_geo_transform = tile.global_geo_transform;
+
+        let model = self.model_file.clone();
+        let ndv = self.no_data_value;
+
+        let predicted_grid = crate::util::spawn_blocking(move || {
+            // extract the actual data from the tiles
+            let pixels: Vec<_> = bot
+                .into_iter()
+                .flat_map(|band| -> Result<Vec<f32>, XGBoostModuleError> {
+                    let mat_tile = band.into_materialized_tile();
+
+                    let res = mat_tile
+                        .grid_array
+                        .inner_grid
+                        .data
+                        .into_iter()
+                        .map(num_traits::AsPrimitive::as_)
+                        .collect::<Vec<f32>>();
+
+                    Ok(res)
+                })
+                .flatten()
+                .collect();
+
+            let model_content =
```

Comment:
try to read the data once in the `query` method once and reuse it for the tiles.

Reply:
Ok, i moved the reading of the file to `_query()` in https://github.com/geo-engine/geoengine/pull/639/commits/a7a59da740eb95fab06e72c3a1a0fa5de0d0cf2c.

I expect, that the cloning of `model_content` is not ok, or is it? I did not find a way without it, that would make the code work, since sooner or later every tile consumes a copy of the model when the booster is created?

Reply:
can you maybe use a `Rc` (or `Arc`) instead of `clone`?

Reply:
`Rc` doesn't work, because `Rc<String>` is not implementing `Send`. With `Arc` it seems to work though.